### PR TITLE
[core]Migrate Gathers operators to new API

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Migrate Gathers operators to new API, remove old includes, functions, namespaces.
 - Refactor to reduce bin-size
 - Minor refactor reference implementation to remove redundant code.

### Tickets:
 - [CVS-126104](https://jira.devtools.intel.com/browse/CVS-126104)

### CODE REVIEW

File: src/core/src/op/gather.cpp [hunk 1]
```diff
@@ -2,42 +2,40 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-#include "openvino/op/gather.hpp"
-
 #include "itt.hpp"
-#include "openvino/core/validation_util.hpp"
+// #include "openvino/core/validation_util.hpp"
```
Comment (t-jankowski) [inline src/core/src/op/gather.cpp:6]: ?

**Review by t-jankowski — APPROVED** [review]

File: src/core/src/op/gather.cpp (hunk 1 above)
Comment (praasz) [reply src/core/src/op/gather.cpp:6]: Removed

File: src/core/src/op/util/gather_base.cpp [hunk 2]
```diff
@@ -5,145 +5,40 @@
...
-                                          out->get_shape(),
-                                          axis,
-                                          batch_dims);
-    } else {
-        OPENVINO_THROW("Unexpected type ", arg1->get_element_type().c_type_string(), " for Gather evaluate method.");
-    }
 
-    return true;
-}
+Shape out_shape_infer(const Shape& data_shape, const Shape& indices_shape, int64_t axis, int64_t batch_dims) {
```
Comment (vurusovs) [inline src/core/src/op/util/gather_base.cpp:22]: why not use `shape_infer()`?

Comment (praasz) [reply src/core/src/op/util/gather_base.cpp:22]: Is just complex as shapes will be converted to PartialShape, there is a lot of check which are not required.
As this implementation worked and shapes are correct to make correct calculations be reference implementation I decide to keep it and re-factored to use `std` functions. 

As future improvement this implementation could be merged with shape_infer to give required validation and simplicity of calculation, but didn't want to change it in this PR

**Pull request merged**